### ISSUE TITLE
AsyncIO Error with Stage 3 + NVME offload

### ISSUE THREAD

**Issue by stephenrawls** [microsoft/DeepSpeed#1070]
Hi,

When trying to use ZERO Stage 3 with NVME offloading, which is required for fitting large models into memory, I am seeing the following error:
```
/nvme/zero_stage_3/fp16params/rank24/0_param.tensor.swp: buffer nbytes != file bytes 28824000 != 28311552
python: /usr/local/lib/python3.6/dist-packages/deepspeed/ops/csrc/aio/py_lib/deepspeed_py_aio_handle.cpp:223: int deepspeed_aio_handle_t::pread(const at::Tensor&, const char*, bool, bool): \
Assertion `static_cast<long long int>(buffer.nbytes()) == num_file_bytes' failed.
```
I have inserted some debug `print()` statements at both the write and read python call sites, here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/swap_tensor/utils.py#L19-L26

And I observe that when I am writing, the tensor really is trying to write `28824000` bytes (because it has `14412000` elements). However when I do an `ls -l /nvme/zero_stage_3/fp16params/rank24/0_param.tensor.swp` I observe that the file only has `28311552` bytes as mentioned in the error message. So it seems that somehow the async write command is failing to properly write the full contents of the tensor.

Any idea why this would happen? Or suggestions for how to debug further?

I have tried looking at kernel logs via `dmesg`, but nothing turns up. I have also tried running program with `strace -e trace=io_submit,io_getevents,io_setup`, but I only see the `io_setup` syscall and not the `io_submit` or `io_getevents` syscalls.

I do have libaio_dev installed.

My deepspeed config looks like this:
```
zero_optimization:
  stage: 3
  stage3_prefetch_bucket_size: 1e9
  stage3_param_persistence_threshold: 1e6
  stage3_max_live_parameters: 1e9
  overlap_comm: true
  contiguous_gradients: true
  offload_param:
      device: nvme
      nvme_path: /nvme
      pin_memory: false
      max_in_cpu: 1e9
      buffer_size: 1e9
      buffer_count: 5
    offload_optimizer:
      device: nvme
      nvme_path: /nvme
      pin_memory: false
```

Thanks,
Stephen

**Comment by tjruwase** [comment]
@stephenrawls , thanks for reporting this issue.  Can you share details for us to repro this issue?

**Comment by tjruwase** [comment]
@stephenrawls, I think this is now fixed. Can you please verify with the master branch? Thanks!

**Comment by stephenrawls** [comment]
I get same error with both latest master branch commit (#ee4deabd75331e9c1bb1383b2068fdebcb33ed08) and also your branch that is mentioned above (#42825bc96217c0969524e449f6901a7d0d72fa5e)

**Comment by tjruwase** [comment]
@stephenrawls, that is strange. Is it possible for you to share your model or model dimensions? That way, I can try a repro on my side?

**Comment by stephenrawls** [comment]
I am using a huggingface RobertaForMaskedLM model with this config:
```
    max_position_embeddings: 512
    type_vocab_size: 1
    num_attention_heads: 40
    num_hidden_layers: 40
    hidden_size: 7680
    intermediate_size: 30720
    gradient_checkpointing: true
```
I chose the size a bit arbitrarily just to test a model that was too big to fit without NVME offloading.

Not sure the easiest way to get you a reproducible script, but I will look into it.

**Comment by tjruwase** [comment]
Can you share the command-line that repros with huggingface RobertaForMaskedLM?

**Comment by stephenrawls** [comment]
I am able to reproduce problem with this script:

```
import torch
from deepspeed.ops.aio import AsyncIOBuilder
from deepspeed.runtime.swap_tensor.constants import AIO_BLOCK_SIZE, AIO_QUEUE_DEPTH, \
    AIO_THREAD_COUNT, AIO_SINGLE_SUBMIT, AIO_OVERLAP_EVENTS


tensor = torch.ones(5000, 5000)

aio_op = AsyncIOBuilder().load(verbose=False)

block_size = 1048576
queue_depth = 8
single_submit = False
overlap_events = True
thread_count = 1
write_hanlde = aio_op.aio_handle(block_size, queue_depth, single_submit, overlap_events, thread_count)

elements_per_buffer = 1e9
param_buffer_count = 5

buffers = torch.empty(int(elements_per_buffer * param_buffer_count),
                      dtype=torch.half, pin_memory=True, requires_grad=False)

buffer = buffers.narrow(0, 0, tensor.numel())

print("starting to write")

write_hanlde.async_pwrite(buffer, "/ssd1/fake-buff.swp")

print("About to wait")
write_hanlde.wait()

print("Done writing")
```

It is supposed to write out `5000 * 5000 * 2 = 50000000 bytes`. And indeed after building DeepSpeed with `#define DEBUG_DS_AIO_PERF 1` I see this in output:
```
starting to write
About to wait
deepspeed_aio: start write 50000000 bytes with 48 io blocks
submit(usec) 4225.485000 io_index=0 nr=8 buf=0x7f9610000000 len=1048576 off=0 
submit(usec) 7744.485000 io_index=8 nr=7 buf=0x7f9610800000 len=1048576 off=8388608 
submit(usec) 9146.070000 io_index=15 nr=7 buf=0x7f9610f00000 len=1048576 off=15728640 
submit(usec) 8854.670000 io_index=22 nr=7 buf=0x7f9611600000 len=1048576 off=23068672 
submit(usec) 9329.830000 io_index=29 nr=7 buf=0x7f9611d00000 len=1048576 off=30408704 
submit(usec) 8993.077000 io_index=36 nr=7 buf=0x7f9612400000 len=1048576 off=37748736 
submit(usec) 4984.491000 io_index=43 nr=5 buf=0x7f9612b00000 len=1048576 off=45088768 
deepspeed_aio: latency statistics(usec) submit min/max/avg = 4225.49 9329.83 7610
deepspeed_aio: latency statistics(usec) complete min/max/avg = 1.258 1068.47 135
deepspeed_aio: runtime(usec) 54681.2 rate(GB/sec) = 0.914391
deepspeed_aio: finish write 50000000 bytes 
```

However my file system shows me the file has wrong number of bytes:
```
$ ls -l /ssd1/fake-buff.swp 
-rw------- 1 root root 49283072 May 18 21:18 /ssd1/fake-buff.swp
```

I have confirmed this is the case after formatting drive with both ext4 and xfs filesystems.

**Comment by stephenrawls** [comment]
Note:

If I set the tensor I am writing above to be a multiple of block-size, e.g.:
```
tensor = torch.ones(8388608)
```

Then I get the expected output size.

My guess is the problem is somehow in the async-io "carry-over" logic when writing out the last block that is less than the block size.

**Comment by stephenrawls** [comment]
One more piece of info:

I added a `for` loop at this line to print debug info for all the IOCBs, instead of just the first one: https://github.com/stephenrawls/DeepSpeed/blob/master/csrc/aio/common/deepspeed_aio_common.cpp#L100-L109

Upon doing that I noticed the following:
```
iocb 0:  submit(usec) 5359.223000 io_index=43 nr=5 buf=0x7f6e22b00000 len=1048576 off=45088768 
iocb 1:  submit(usec) 5359.223000 io_index=43 nr=5 buf=0x7f6e22c00000 len=1048576 off=46137344 
iocb 2:  submit(usec) 5359.223000 io_index=43 nr=5 buf=0x7f6e22d00000 len=1048576 off=47185920 
iocb 3:  submit(usec) 5359.223000 io_index=43 nr=5 buf=0x7f6e22e00000 len=1048576 off=48234496 
iocb 4:  submit(usec) 5359.223000 io_index=43 nr=5 buf=0x7f6e22f00000 len=716928 off=49283072 
```
The discrapancy in file size is exactly `716928` bytes, meaning as suspected the "carry-over" buffer writing the last remainder out to disk is what failed to write.

If I modify my buffer size to be something that is divisible by 512 bytes (the SSD devices block size), then everything works fine. i.e.:
```
tensor = torch.ones(6331392)
```
leads to 
```
...
iocb 0:  submit(usec) 6186.319000 io_index=8 nr=5 buf=0x7f9e58800000 len=1048576 off=8388608 
iocb 1:  submit(usec) 6186.319000 io_index=8 nr=5 buf=0x7f9e58900000 len=1048576 off=9437184 
iocb 2:  submit(usec) 6186.319000 io_index=8 nr=5 buf=0x7f9e58a00000 len=1048576 off=10485760 
iocb 3:  submit(usec) 6186.319000 io_index=8 nr=5 buf=0x7f9e58b00000 len=1048576 off=11534336 
iocb 4:  submit(usec) 6186.319000 io_index=8 nr=5 buf=0x7f9e58c00000 len=79872 off=12582912 
...
deepspeed_aio: finish write 12662784 bytes 
Done writing
```
And also the file has expected number of bytes:
```
ls -l /ssd1/fake-buff.swp 
-rw------- 1 root root 12662784 May 18 23:22 /ssd1/fake-buff.swp
```

Notice that the last "carry-over" write has size `79872` bytes, which is divisible by `512`.

It seems that block-level alignment is important when dealing with O_DIRECT flag on files: http://manpath.be/f33/2/open#L843

**Comment by tjruwase** [comment]
@stephenrawls, thanks for sharing your observations. As you have rightly observed the aio library does not support writes that are not aligned to block size of 512. This is expected given the following considerations. 

1. The original issue identified a real problem where parameter offloading in ZeRO-Infinity was not handling the misalignment issue and this PR #1076 has fixed that. Can you please verify that you no longer see the original error report of parameter offloading when using ZeRO-Infinity.

2. The aio library is not currently meant to be used directly but rather through clients that avoid this misalignment problem. As you have observed ensuring that the write size is aligned to block size works correctly.
 
3. Do you currently have a compelling use of the aio library outside of ZeRO-Infinity? Perhaps we can discuss that usage scenario somehow. Keep in mind introducing padding allocation to handle alignment issues could hurt the lightweight nature and performance. You might find #998 useful.

To reiterate, can you please confirm that #1076 fixes the original issue regarding parameter offloading?

**Comment by stephenrawls** [comment]
Ah, I see the PR you mentioned was just merged yesterday. I tried again, using DeepSpeed commit d88d92799553cb75af536fd4af766ec56e4018cd.

But I still see the same problem:
```
/ssd1/zero_stage_3/fp16params/rank26/0_param.tensor.swp: buffer nbytes != file bytes 57648000 != 57648128
python: /usr/local/lib/python3.6/dist-packages/deepspeed/ops/csrc/aio/py_lib/deepspeed_py_aio_handle.cpp:223: int deepspeed_aio_handle_t::pread(const at::Tensor&, const char*, bool, bool): Assertion `static_cast<long long int>(buffer.nbytes()) == num_file_bytes' failed.
```

> Do you currently have a compelling use of the aio library outside of ZeRO-Infinity?

I do not. I simply tried to use it to create a minimally-reproducing script that demonstrates the bug I am seeing, because running the full training code takes a long time until it reports an error.

**Comment by stephenrawls** [comment]
@tjruwase  I have identified the problem is here: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage3.py#L1027-L1029

Basically all the aio read calls directly in `partitioned_param_swapper.py` are properly aligned after the PR you pointed me to, because the swap-in buffers are explicitly aligned.

However in the above code I pointed to, there is still a code-path that directly calls the `swap_in()` function with an explicit (and un-aligned) `swap_in_buffers` parameter.

**Comment by stephenrawls** [comment]
Actually, everything works fine if I use a single GPU. I think the underlying problem is this line:
https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/partition_parameters.py#L681-L682

i.e.:
```
tensor_size = self._aligned_size(param)
partition_size = tensor_size // self.world_size
```
While the tensor-size is aligned properly, after dividing by world size the `partition_size` is not aligned.

And the `get_buffer()` call here uses the `compute_buffer` not the aligned `swap_buffer`: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/partition_parameters.py#L689

**Comment by tjruwase** [comment]
@stephenrawls, I am absolutely blown away by your way, way, way, above-and-beyond efforts in isolating this issue. I wont be able to fix this until later. But I don't want you to spend more cycles on this, as you have provided more than sufficient information to fix it. On the other hand, if you want to submit a PR, I will be glad to guide that process. That way you get the full credit that you deserve for resolving this issue.

**Comment by stephenrawls** [comment]
Thanks, that's kind of you to say.

If you think fix on your end will be more than a couple days, then I can find time to work on fix + PR. Do you have any quick pointers on what is cleanest way to make this flat-buffer use the aligned parameter sizes? If so I can run with that and work on implementation.

**Comment by tjruwase** [comment]
@stephenrawls, actually I do have time to handle this. After a closer look it appears that there are some other issues. For exampe, this [line ](https://github.com/microsoft/DeepSpeed/blob/c1550b89ba796d89a10f0445d896896d9206bf26/deepspeed/runtime/swap_tensor/partitioned_param_swapper.py#L302) will fail after the current bug is fixed, because `compute_buffers ` is not defined on that code path. I will keep you updated.

**Comment by tjruwase** [comment]
> And the `get_buffer()` call here uses the `compute_buffer` not the aligned `swap_buffer`: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/partition_parameters.py#L689

This behavior is actually fine because the idea is to hide the alignment issues from logic outside `swap_tensor/partitioned_param_swapper.py`. When the param is submitted for a swap operation, the swapper knows how to replace with the corresponding aligned swap_buffer using `self.param_id_to_swap_buffer`

**Comment by stephenrawls** [comment]
Yup, I hit that bug w.r.t. `compute_buffers` yesterday after trying a fix.

Okay sounds like you have it under control then, thanks for looking into it.

**Comment by stephenrawls** [comment]
@tjruwase yes, I have tested #1086 via commit 38d46848f450a080c8ab96427d9da00c2e5b6327, and it works for me now. Thanks for quick fix!
